java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/ContextAdapter.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight.grpc;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.arrow.flight.CallContext;
+import org.apache.arrow.flight.auth.AuthConstants;
+
+import io.grpc.Context;
+
+/**
+ * Adapter class for gRPC contexts.
+ */
+public class ContextAdapter implements CallContext {
+  // gRPC uses reference equality when looking up keys in a Context. Cache used keys in this static map
+  // so that look ups can succeed.
+  private static final Map<String, Context.Key<String>> usedKeys = new HashMap<>();
```

Comment:
An internal synchronized cache may be surprising...I'd say if we have to have a cache, it should just use ConcurrentHashMap, and preferably, we should just wrap Context.Key<T> ourselves.

Reply:
Yeah I wasn't really happy about this.

I'm not sure what you mean by wrapping Context.Key here. The issue is that Context.key("foo") != Context.key("foo") when invoking key.getValue() since gRPC uses reference equality for context keys. We need to reference the _same_ Context.key when writing the value in the middleware and getting the value to set the peer identity before executing Flight calls. However neither side should be aware that they are using gRPC Contexts so we can't a shared static value representing the key as we did before these changes.

Does ConcurrentHashMap allow for atomic writes of the same key if you use putIfAbsent?

Reply:
You could implement a 
```java
class ContextKey<T> {
    private final Context.Key<T> key;
}
```
and give it the same restriction (instances are compared by identity, not value), right?

Reply:
Also of note, this should at least be a ConcurrentMap since it's static and shared by multiple requests.

Reply:
Thinking about it, would a better implementation be to store a `Map<String, String>` implementing RequestContext inside the gRPC context, and just pull it out of the context wherever needed? That way you don't have to deal with gRPC context keys. (I do like the context key pattern because it is typed, though!)

Reply:
I've implemented the use of Map<String,String> as the RequestContext now. This is much better, great suggestion!